contracts/clients-tests/BlsProvider.test.ts

```diff
@@ -0,0 +1,514 @@
+      Error,
+      "Call provider.getSigner first.",
+    );
+  });
+
+  it("should return failures as a json string and throw an error when sending an invalid transaction", async () => {
+    // Arrange
+    const invalidEthValue = parseEther("-1");
+
+    const unsignedTransaction = {
+      value: invalidEthValue,
+      to: signers[1].address,
+      data: "0x",
+    };
+    const signedTransaction = await blsSigner.signTransaction(
+      unsignedTransaction,
+    );
+
+    // Act
+    const result = async () =>
+      await blsProvider.sendTransaction(signedTransaction);
+
+    // Assert
+    await expect(result()).to.be.rejectedWith(
+      Error,
+      '[{"type":"invalid-format","description":"field operations: element 0: field actions: element 0: field ethValue: hex string: missing 0x prefix"},{"type":"invalid-format","description":"field operations: element 0: field actions: element 0: field ethValue: hex string: incorrect byte length: 8.5"}]',
+    );
+  });
+
+  it("should throw an error when the transaction receipt cannot be found", async () => {
+    // Arrange
+    const invalidTransactionHash = id("invalid hash");
+    const retries = 1; // Setting this to 1 as we do not to wait in order for the logic to be correctly tested
+
+    // Act
+    const result = async () =>
+      await blsProvider._getTransactionReceipt(
+        invalidTransactionHash,
+        1,
+        retries,
+      );
+
+    // Assert
+    await expect(result()).to.be.rejectedWith(
+      Error,
+      `Could not find bundle receipt for transaction hash: ${invalidTransactionHash}.`,
+    );
+  });
+
+  it("should wait for a transaction and resolve once transaction hash is included in the block", async () => {
+    // Arrange
+    const recipient = signers[1].address;
+    const transactionResponse = await blsSigner.sendTransaction({
+      to: recipient,
+      value: parseEther("1"),
+    });
+
+    // Act
+    const transactionReceipt = await blsProvider.waitForTransaction(
+      transactionResponse.hash,
+      1,
+      10,
+    );
+
+    // Assert
+    // TODO: bls-wallet #412 Update values returned in bundle receipt to more closely match ethers transaction response
+    expect(transactionReceipt).to.be.an("object").that.deep.includes({
+      to: "0x",
+      from: "0x",
+      contractAddress: "0x",
+      // gasUsed: BigNumber.from("0"),
+      logsBloom: "",
+      logs: [],
+      confirmations: transactionResponse.confirmations,
+      // cumulativeGasUsed: parseEther("0"),
+      // effectiveGasPrice: parseEther("0"),
+      byzantium: false,
+      type: 2,
+    });
+
+    expect(transactionReceipt).to.include.keys(
+      "transactionIndex",
+      "blockHash",
+      "transactionHash",
+      "blockNumber",
+    );
+
+    // TODO: parseEther() doesn't work when asserting this way ^.
+    // Expects BigNumber { value: "0" } from transactionReceipt,
+    // but this returns { _hex: '0x00', _isBigNumber: true }.
```

Comment:
Got this weird issue where my assertions do not like `parseEther` and `BigNumber.from` now. The tests pass with the below assertions (lines 337-339 at the time of writing), but they do not work with the above assertion using `.includes`. Not sure why this is happening?

To test this, uncomment the `gasUsed`, `cumulativeGasUsed` or `effectiveGasPrice` values above.

Reply:
That makes sense. Each `BigNumber` (BN) is a separate object instance, meaning two BNs with the same value will not equal, hence why `.eql` is used instead in most places in code with BNs.

Primitives in JavaScript will equal each other, but more complex types like arrays and objects will not. See https://stackoverflow.com/a/359509 for some zany ways they will with `==`. (general best practice is to always use `===`)

```typescript
const a = { a: 1 }
const b = { a: 1 }
a === b // false
a === a // true
a.a === b.a // true

BigNumber.from("1") === BigNumber.from("1") // false
BigNumber.from("1").eql(BigNumber.from("1")) // true
BigNumber.from("1").toNumber() === 1 // true
```

`.includes` should use strict equality (`===`) which means BNs would fail that check.

To resolve this, you can:
- Do your expects outside the `.include` (what you have done)
- Convert the txn receipt props to primitive types (string, number)
- Use a chai pulgin such as [chai-bignumber](https://www.chaijs.com/plugins/chai-bignumber/). `ethers.js` uses its own BN lib so this may not work.

Reply:
Ahh yeah that part makes a lot more sense now. The one part that still confuses me is how `expect(transactionReceipt.gasUsed).to.equal(parseEther("0"));` still passes. 

According to the chai docs, `.equal` "Asserts that the target is strictly (===) equal to the given val." So if BNs fail using strict equality, that assertion shouldn't pass? I tested for a false positive with `expect(transactionReceipt.gasUsed).to.equal(parseEther("1"));` as a sanity check, and that assertion fails so confused how the first assertion passes. Am I understanding correctly?

Reply:
In terms of what I went for in the end:

Tried `chai-bignumber` to no avail. 

Decided against converting the receipt properties to primitive types, as that would mean diverging from the ethers return type of `TransactionReceipt`. 

Decided to go with option 1 and leave the assertions as they are. Think it makes more sense to change the test code to work with the implementation code, rather than change the implementation code just to make the tests pass in a cleaner way.